### ISSUE TITLE
Add input field to Quick Menu (feature suggestion)

### ISSUE THREAD

**Issue by tzeumer** [ssborbis/ContextSearch-web-ext#41]
Thanks for this great add-on. When [Web Search Pro](https://addons.mozilla.org/en-US/firefox/addon/web-search-pro/) extension died with Quantum I really missed the second search field and the context sensitive search. I even switched from Firefox to the Chrome derivative Vivaldi, because the [SearchBar](https://chrome.google.com/webstore/detail/searchbar/fjefgkhmchopegjeicnblodnidbammed) extension provided something similar.

With SearchBar I came to really like its feature of an input field in which the selected text is copied and can be edited prior to searching. Often I just select some random text, because it's such a quick and convenient way to search with any search engine. 

It would be great for ContextSearch to have such a field between the tools and the icons in the quick menu. Maybe in addition the topmost engine could be used as default when pressing enter.

Pushing the idea a bit further, it would be nice if the extension button would open with a Quick Menu tab (right before the Menus tab) - this way one nearly would have a second search field with probably another default search engine than the Firefox search field.

**Comment by ssborbis** [comment]
I've got the basics down.  I added a title bar of sorts to the quick menu that displays the current search terms. Click it and it becomes an input field.

**Comment by ssborbis** [comment]
Aesthetics. Centered text looks better when you're not editing the terms, but longer terms require left alignment to prevent some text from being hidden. I figured you would probably want the input being focused on launch so I'm making it an option.

**Comment by tzeumer** [comment]
Ah, this is a good reason. Maybe it could work this way: It's always a title bar, but onKeyDown it just becomes an input field. At least I can't think of a case where on would select text and type somewhere else in the browser while the Quick Menu is open.

**Comment by ssborbis** [comment]
Good idea. I'll add the listener

**Comment by tzeumer** [comment]
After thinking a minute, I thought of a case. If the menu pops up in an input or textarea field you can just type on currently. Maybe the listener should only work if the focus is not in one of these fields (then you just would have to click the titlebar).

**Comment by ssborbis** [comment]
maybe use a specific key to listen for like Tab to focus input and select all text making it readily replaced vs appended. I'll play around with some settings

**Comment by tzeumer** [comment]
I think it is a very special case that only "might" bother people who use the auto popup and have the option for input/textarea enabled (which is my use case, too). I think, ignoring the onKeyDown for just these cases would be a good compromise between "intuitive" usability and not having to add an extra option. 

But I think there might be a wide spread between how people prefer to use the menu and so it's never easy to tell what's best. Thanks a lot for considering at all :)

**Comment by ssborbis** [comment]
After some testing, it works well just setting focus on quick menu open with left-aligned text. I think I might have needed to do some background messaging to capture keydowns from the parent window anyway so this simplifies things and still looks good. The search bar can be turned off via options. I'm updating the repo as I go so feel free to grab the src , load up as a temp addon and see what you think

**Comment by tzeumer** [comment]
This looks and works great! I can't find a possible way there could be any trouble using it; works well with Input and Textarea, too. I like it very much :)

Maybe there could a little bit more padding top/bottom from the text and the contrast white/blue might be a little too low for some people. But that are just details and I'm just guessing. Overall it's cool.

Off topic: I guess the things in the Screenshot (minimal vertical scrollbar when viewing the options via button and two "undefined" options at the bottom) are just work in progress? :)

![2018-05-27_181214-contextsearch](https://user-images.githubusercontent.com/119825/40588354-8466b4a6-61dc-11e8-9387-38ea72745c28.png)

**Comment by ssborbis** [comment]
Yep, just a test area for some code. It'll be gone soon. The horizontal scroll a css glitch from the wider divs required for the german translation. I'll get around to it sometime

**Comment by ssborbis** [comment]
search bar live in v1.5.6

**Comment by tzeumer** [comment]
Great. Looks good, works like a charm. Thank you! :)

**Comment by ssborbis** [comment]
> Pushing the idea a bit further, it would be nice if the extension button would open with a Quick Menu tab (right before the Menus tab) - this way one nearly would have a second search field with probably another default search engine than the Firefox search field.

I've converted the browser action button to mimic Firefox's search toolbar. The current config shows all search engines, hidden as well, in the same format as the quick menu. Works pretty well. I've also added the ability in both the quick menu and the browser action menu to click a search engine with an empty string, taking the user to the search engine website without running a query. Let me know what you think

**Comment by hg42** [comment]
wow, that's really useful, thanks!

Now this creates new wishes :-)

When typing something, it would be nice to select a search engine with the keyboard.
Intuitively I tried the tab key and would have expected to scan through the buttons with further tabs.
Alternatively I tried cursor down (or may be up when the input field is at the bottom, btw. does anyone use this?) and would have expected to move around with the arrow keys (and wrapping at the last or first button in a row).

**Comment by ssborbis** [comment]
I've got the tab working to cycle through search engines. i might leave it at that. The more pressing problem is this 

#45

**Comment by hg42** [comment]
even faster :-) thanks

**Comment by tzeumer** [comment]
Just gone for a few hour and already a new version. I like it very much. The status bar in the button version is nice - it's a good way too quickly distinguish between identical icons (for example "Wikipedia DE" and "Wikipedia EN").  Maybe it could be added to the "inline" QuickMenu too? Maybe it would also serve as a nice visual separator between the tools and the engine icons? But it's really just a thought and nothing I'm really missing :)

I think all new strings are translated too already. Only one is missing for german. Maybe it could be added ad hoc :)
`"hidden": {"message": "Ausgeblendet"},`

Well, I think this concludes this ticket in a best time? :)

**Comment by ssborbis** [comment]
>  Often I just select some random text, because it's such a quick and convenient way to search with any search engine.

I added the optional hotkey Ctrl-Q (disabled by default) to bring up the quick menu anytime with focus set on the empty search bar. You're probably used to selecting random text, but that option is there now if you want it.

> I think all new strings are translated too already. Only one is missing for german. Maybe it could be added ad hoc :)
"hidden": {"message": "Ausgeblendet"},

There are a few more strings now in 1.5.8. It's probably best I don't try to translate them. The literal conversion is probably wrong.

**Comment by tzeumer** [comment]
> I added the optional hotkey Ctrl-Q (disabled by default) to bring up the quick menu anytime with focus set on the empty search bar. You're probably used to selecting random text, but that option is there now if you want it.

Old habits are hard to get rid of, but it's a good feature nonetheless. On the setting side the hotkey method seems a bit redundant to the keyboard method (the only difference is marked text, no marked text). I don't really mind, but wondered anyway, if extension icons can have a hotkey in Firefox. If the hotkey would open the extension menu, you get practically the same result, but it would even work on pages like addons.mozilla.org.

> There are a few more strings now in 1.5.8. It's probably best I don't try to translate them. The literal conversion is probably wrong.

I  created a pull request. I hope I didn't miss anything.

**Comment by ssborbis** [comment]
I had considered using the hotkey for the toolbar action. It's good to get some user input, I'm a little too deep in the woods to be objective. 

Merging changes... Thanks for the update.

**Comment by tzeumer** [comment]
Probably there is no objective best and it's often more a matter of subjective preferences. So, maybe for some design choices it's the best to wait until more users ask for a specific change :)
